HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+
+```
+hooks.evm.pureIntrinsicGasCost=0
+hooks.evm.lambdaIntrinsicGasCost=0
+```
+
+### Throttling
+
+We propose EVM hooks be subject to the same gas throttle as top-level contract calls. Specifically, when an EVM hook
+executes, its initial EVM sender address is the payer of the referencing transaction. If this payer is throttle exempt,
+no throttles are applied. Otherwise, if the network is at capacity for gas usage, EVM hook execution will be throttled
+on that basis and the triggering transaction fail with a status of `EVM_HOOK_EXECUTION_THROTTLED`.
+
+### EVM hook execution
+
+The first message frame for the EVM transaction that executes an EVM hook will be a logical `DELEGATECALL` from the
+**hook system contract address `0x16d`** to the hook's implementing `ContractID`, sent by the payer of the upfront gas
+cost for the hook's execution.
+
+As a concrete example, suppose account `0.0.A` installs an EVM lambda hook for the `ACCOUNT_ALLOWANCE_HOOK` extension
+point at index `1`. The hook's implementing contract is `0.0.H` with EVM address `0xab...cd`. Now `0.0.B` with EVM
+address `0x01...23` sends a `CryptoTransfer` transaction that references the hook `0.0.A.1` with gas limit `100_000`.
+
+The network will construct an initial EVM message frame with,
+1. `sender` address `0x01...23`;
+2. `receiver` address `0x16d`;
+3. `contract` address `0xab...cd` (hence the source of the executing bytecode);
+4. Storage of the `0.0.A.1` lambda EVM hook; and,
+5. Gas remaining of `100_000` (no intrinsic gas cost).
+
+Since this hook is a lambda, it can then proceed with arbitrary EVM behaviors, including calls to external oracles,
+`SLOAD` and `SSTORE` operations with its storage, and so on.
+
+If the `0.0.A.1` hook is a pure EVM hook, it executes in an extremely restricted mode. Not only is the initial frame
+marked **static**, prohibiting all state-changing operations; but the network also disables the `PREVRANDAO`, `SLOAD`,
+and `CALL` opcodes. That is, a pure hook cannot do anything but apply a pure function to its input data.
+
+Although users will doubtless uncover a very wide range of use cases for EVM hooks, we expect the single most common
+type of hook contract to implement a single external method that reverts when not executed by the network _as a hook_.
+That is,
+
+```solidity
+contract HookContract {
+    function hookFunction(address installer) external payable {
+        // Revert if we are not executing a DELEGATECALL from the hook system contract
+        require(address(this) == 0x16d, "Contract can only be called as a hook");
+        // Continue executing as a hook on behalf of the installer address
+    }
+}
+```
+
+Because a hook is conceptually a function of the installing account, when it calls a Hedera system contract, the system 
+contract will see that account's address as the sender. If a hook calls another contract, it is valid for that contract 
```

Comment:
> Because a hook is conceptually a function of the installing account, when it calls a Hedera system contract, the system contract will see that account's address as the sender.

It seems very risky to execute with the account privilege something triggered by any external sources.
I suggest to consider the hook as a normal smart contract, using their own privileges completely separated by the caller/installer's ones. This way, a system contract will see the sender as the hook's address and not as the hook installer address.

Reply:
A valid concern; as emphasized [here](https://github.com/tinker-michaelj/hiero-improvement-proposals/blob/main/HIP/hip-1195.md#the-evm-environment-for-hooks) and discussed extensively in last two TSC calls, we expect hooks to be introduced through application HIPs that go through extensive community auditing and ecosystem adoption.